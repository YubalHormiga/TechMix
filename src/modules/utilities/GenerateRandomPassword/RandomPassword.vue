<script setup>
import { ref, inject } from 'vue'

const passwordBox = ref('')
const toast = inject('toast')
const createPassword = () => {
  passwordBox.value = crypto.randomUUID()
}

const copyPassword = async () => {
  try {
    await navigator.clipboard.writeText(passwordBox.value)

    toast.open({
      message: 'Contrase単a copiada al portapapeles',
      type: 'info'
    })
  } catch (err) {
    console.error('Error al copiar la contrase単a: ', err)
  }
}
</script>

<template>
  <div class="flex flex-col justify-center items-center mt-10">
    <div
      class="flex flex-col items-center justify-center w-[50rem] h-60 rounded-xl text-[#1b933b] bg-[#060a07]"
    >
      <div class="mb-3">
        <h2 class="text-4xl text-[#eff1ef]">Crear Contrase単a Aleatoria</h2>
      </div>
      <div class="flex justify-center gap-3 items-center w-full h-15">
        <div>
          <input type="text" :value="passwordBox" class="p-1 w-[33rem] text-center" readonly />
        </div>
        <div>
          <svg
            xmlns="http://www.w3.org/2000/svg"
            width="40"
            height="40"
            viewBox="0 0 24 24"
            @click="copyPassword"
            class="cursor-pointer"
          >
            <g
              fill="none"
              stroke="#eab308"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            >
              <path d="M16 3H4v13" />
              <path d="M8 7h12v12a2 2 0 0 1-2 2h-8a2 2 0 0 1-2-2z" />
            </g>
          </svg>
        </div>
      </div>
      <button @click="createPassword" class="p-3 mt-3 text-[#1e2733] bg-[#95daa7] rounded-lg">
        Generar Contrase単a
      </button>
    </div>
  </div>
</template>

<style scoped></style>
