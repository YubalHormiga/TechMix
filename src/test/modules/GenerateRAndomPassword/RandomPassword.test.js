import { describe, it, expect, vi } from 'vitest'
import RandomPassword from '@/modules/GenerateRandomPassword/RandomPassword.vue'
import { mount } from '@vue/test-utils'

//Component Rendering Text
describe('RandomPassword', () => {
  const wrapper = mount(RandomPassword)
  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('reders correctly', () => {
    expect(wrapper.find('h2').text()).toContain('Crear Contraseña Aleatoria')
    expect(wrapper.find('button').text()).toBe('Generar Contraseña')
  })
})

// Password Generation Test

describe('Password Generator', () => {
  const wrapper = mount(RandomPassword)

  it('generates a password when clicking the button', async () => {
    const button = wrapper.find('button')
    await button.trigger('click')
    const input = wrapper.find('input')
    expect(input.element.value).toHaveLength(36)
  })
})

//Clipboard Copy Test

describe('PasswordGenerator', () => {
  it('Copies de password to the clipboard', async () => {
    const wrapper = mount(RandomPassword)

    const button = wrapper.find('button')
    await button.trigger('click')

    Object.assign(navigator, {
      clipboard: {
        writeText: vi.fn(() => Promise.resolve())
      }
    })
    const copyIcon = wrapper.find('svg')
    await copyIcon.trigger('click')
    expect(navigator.clipboard.writeText).toHaveBeenCalledWith(wrapper.vm.passwordBox)
  })
})

describe('PasswordGenerator', () => {
    it('shows an error if unable to copy password', async () => {
      const wrapper = mount(RandomPassword)
  
      Object.assign(navigator, {
        clipboard: {
          writeText: vi.fn(() => Promise.reject('Copy Error')) 
        }
      })
  
      const consoleSpy = vi.spyOn(console, 'error')
  
      const button = wrapper.find('button')
      await button.trigger('click')
  
      const copyIcon = wrapper.find('svg')
      await copyIcon.trigger('click')
  
      expect(consoleSpy).toHaveBeenCalledWith('Error al copiar la contraseña: ', 'Copy Error')
  
      consoleSpy.mockRestore()
    })
  })
  