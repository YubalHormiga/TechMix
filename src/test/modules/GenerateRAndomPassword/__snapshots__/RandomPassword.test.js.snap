// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RandomPassword > Coincidir con la instacia 1`] = `
"<div class="flex flex-col justify-center items-center mt-10">
  <div class="flex flex-col items-center justify-center w-[50rem] h-60 rounded-xl text-[#1b933b] bg-[#060a07]">
    <div class="mb-3">
      <h2 class="text-4xl text-[#eff1ef]">Crear Contraseña Aleatoria</h2>
    </div>
    <div class="flex justify-center gap-3 items-center w-full h-15">
      <div class=""><input type="text" class="p-1 w-[33rem] text-center" readonly="" value=""></div>
      <div><svg xmlns="http://www.w3.org/2000/svg" width="40" height="40" viewBox="0 0 24 24" class="cursor-pointer">
          <g fill="none" stroke="#eab308" stroke-linecap="round" stroke-linejoin="round" stroke-width="2">
            <path d="M16 3H4v13"></path>
            <path d="M8 7h12v12a2 2 0 0 1-2 2h-8a2 2 0 0 1-2-2z"></path>
          </g>
        </svg></div>
    </div><button class="p-3 mt-3 text-[#1e2733] bg-[#95daa7] rounded-lg">Generar Contraseña</button>
  </div>
</div>"
`;

exports[`RandomPassword > Se renderiza correctamente 1`] = `
"<div class="flex flex-col justify-center items-center mt-10">
  <div class="flex flex-col items-center justify-center w-[50rem] h-60 rounded-xl text-[#1b933b] bg-[#060a07]">
    <div class="mb-3">
      <h2 class="text-4xl text-[#eff1ef]">Crear Contraseña Aleatoria</h2>
    </div>
    <div class="flex justify-center gap-3 items-center w-full h-15">
      <div class=""><input type="text" class="p-1 w-[33rem] text-center" readonly="" value=""></div>
      <div><svg xmlns="http://www.w3.org/2000/svg" width="40" height="40" viewBox="0 0 24 24" class="cursor-pointer">
          <g fill="none" stroke="#eab308" stroke-linecap="round" stroke-linejoin="round" stroke-width="2">
            <path d="M16 3H4v13"></path>
            <path d="M8 7h12v12a2 2 0 0 1-2 2h-8a2 2 0 0 1-2-2z"></path>
          </g>
        </svg></div>
    </div><button class="p-3 mt-3 text-[#1e2733] bg-[#95daa7] rounded-lg">Generar Contraseña</button>
  </div>
</div>"
`;

exports[`RandomPassword > matches the snapshot 1`] = `
"<div class="flex flex-col justify-center items-center mt-10">
  <div class="flex flex-col items-center justify-center w-[50rem] h-60 rounded-xl text-[#1b933b] bg-[#060a07]">
    <div class="mb-3">
      <h2 class="text-4xl text-[#eff1ef]">Crear Contraseña Aleatoria</h2>
    </div>
    <div class="flex justify-center gap-3 items-center w-full h-15">
      <div class=""><input type="text" class="p-1 w-[33rem] text-center" readonly="" value=""></div>
      <div><svg xmlns="http://www.w3.org/2000/svg" width="40" height="40" viewBox="0 0 24 24" class="cursor-pointer">
          <g fill="none" stroke="#eab308" stroke-linecap="round" stroke-linejoin="round" stroke-width="2">
            <path d="M16 3H4v13"></path>
            <path d="M8 7h12v12a2 2 0 0 1-2 2h-8a2 2 0 0 1-2-2z"></path>
          </g>
        </svg></div>
    </div><button class="p-3 mt-3 text-[#1e2733] bg-[#95daa7] rounded-lg">Generar Contraseña</button>
  </div>
</div>"
`;

exports[`RandomPassword > renders correctly 1`] = `
"<div class="flex flex-col justify-center items-center mt-10">
  <div class="flex flex-col items-center justify-center w-[50rem] h-60 rounded-xl text-[#1b933b] bg-[#060a07]">
    <div class="mb-3">
      <h2 class="text-4xl text-[#eff1ef]">Crear Contraseña Aleatoria</h2>
    </div>
    <div class="flex justify-center gap-3 items-center w-full h-15">
      <div class=""><input type="text" class="p-1 w-[33rem] text-center" readonly="" value=""></div>
      <div><svg xmlns="http://www.w3.org/2000/svg" width="40" height="40" viewBox="0 0 24 24" class="cursor-pointer">
          <g fill="none" stroke="#eab308" stroke-linecap="round" stroke-linejoin="round" stroke-width="2">
            <path d="M16 3H4v13"></path>
            <path d="M8 7h12v12a2 2 0 0 1-2 2h-8a2 2 0 0 1-2-2z"></path>
          </g>
        </svg></div>
    </div><button class="p-3 mt-3 text-[#1e2733] bg-[#95daa7] rounded-lg">Generar Contraseña</button>
  </div>
</div>"
`;
